File: lsd-interceptors/src/main/java/io/lsdconsulting/interceptors/messaging/EventConsumerInterceptor.java
package io.lsdconsulting.interceptors.messaging;

import com.lsd.LsdContext;
import com.lsd.diagram.ValidComponentName;
import lombok.RequiredArgsConstructor;
import org.springframework.messaging.Message;
import org.springframework.messaging.MessageChannel;
import org.springframework.messaging.support.ChannelInterceptor;

import static io.lsdconsulting.interceptors.http.common.Headers.HeaderKeys.SOURCE_NAME;
import static io.lsdconsulting.interceptors.http.common.Headers.HeaderKeys.TARGET_NAME;
import static io.lsdconsulting.interceptors.messaging.PayloadRetriever.getPayload;

@RequiredArgsConstructor
public class EventConsumerInterceptor implements ChannelInterceptor {

    private final LsdContext lsdContext;

    @Override
    public Message<?> preSend(Message<?> message, MessageChannel channel) {
        String payload = getPayload(message);
        String source = (String) message.getHeaders().get(SOURCE_NAME.key());
        String target = (String) message.getHeaders().get(TARGET_NAME.key());

        lsdContext.capture("Consume event from " + ValidComponentName.of(source) + " to " + ValidComponentName.of(target), payload);

        return message;
    }
}

File: lsd-interceptors/src/main/kotlin/io/lsdconsulting/interceptors/messaging/EventConsumerInterceptor.kt
package io.lsdconsulting.interceptors.messaging

import com.lsd.core.LsdContext
import com.lsd.core.builders.MessageBuilder.Companion.messageBuilder
import com.lsd.core.domain.MessageType
import io.lsdconsulting.interceptors.common.HeaderKeys.TARGET_NAME
import lsd.format.prettyPrint
import org.springframework.beans.factory.annotation.Value
import org.springframework.messaging.Message
import org.springframework.messaging.MessageChannel
import org.springframework.messaging.support.ChannelInterceptor

class EventConsumerInterceptor(
    private val lsdContext: LsdContext,
    @Value("\${info.app.name}") private var appName: String
) : ChannelInterceptor {

    override fun preSend(message: Message<*>, channel: MessageChannel): Message<*> {
        val payload = prettyPrint(message.payload)
        val source =prettyPrint(message.headers[TARGET_NAME.key()])
        val target = appName
        lsdContext.capture(
            messageBuilder()
                .id(lsdContext.idGenerator.next())
                .from(source)
                .to(target)
                .label("Consume event")
                .data(renderHtmlFor(message.headers, payload))
                .type(MessageType.ASYNCHRONOUS)
                .build()
        )
        return message
    }
}
